src/garage/_primitives.py

```diff
@@ -0,0 +1,152 @@
+"""Policy interface."""
+import abc
+
+
+class Primitive(abc.ABC):
+    """Primitive interface.
+
+    Args:
+        name (str): Name for logs, variable scopes, etc.
+        spec (garage.InOutSpec): Input/output specification this primitive will
+            satisfy.
+
+    """
+
+    def __init__(self, name, spec):
+        self._name = name
+        self._spec = spec
+
+    @abc.abstractmethod
+    def get_parameters(self):
+        """Get parameters of this primitive.
+
+        Returns:
+            dict[str, np.ndarray]: A dictionary mapping from names to values of
+                each parameter.
+
+        """
+
+    @abc.abstractmethod
+    def set_parameters(self, parameters):
+        """Set the parameters of this primtive.
+
+        Args:
+            parameters (dict[str, np.ndarray]): A dictionary mapping from names
+                to values of each parameter.
+
+        """
+
+    @property
+    def spec(self):
+        """InOutSpec: The input/output spec for this primitive."""
+        return self._spec
+
+    @property
+    def name(self):
+        """str: Name of this primitive."""
+        return self._name
+
+    @abc.abstractmethod
+    def clone(self, name):
+        """Create a fully-independent copy of this primitive.
+
+        Args:
+            name (str): Name for the cloned instance.
+
+        Returns:
+            Primitive: A fully-independent copy of this primitive.
+
+        """
+
+    @abc.abstractmethod
+    def __getstate__(self):
+        """See Object.__getstate__.
+
+        Returns:
+            dict: Pickling state for the primitive
+
+        """
+
+    @abc.abstractmethod
+    def __setstate__(self, state):
+        """See Object.__setstate__.
+
+        Args:
+            state (dict): Unpickled state of this primitive.
+
+        """
+
+
+class Policy(Primitive):
+    """Policy interface.
+
+    Args:
+        name (str): Name for logs, variable scopes, etc.
+        env_spec (garage.envs.EnvSpec): Environment spec this Policy
+            will satisfy.
+
+    """
+
+    def __init__(self, name, env_spec):
+        super().__init__(name, env_spec.transpose())
+
+    @abc.abstractmethod
+    def get_action(self, observation, policy_state=None):
```

Comment:
Why can `policy_state` be `None`, but I can also explicitly request a reset state? What's the semantic difference between passing `None` and passing the result of `get_reset_state`.

Reply:
Does this `policy_state` apply to only recurrent policies?

Reply:
@krzentner This allows for calling the `get_action()` API without a `policy_state`, which is idiomatic for non-stateful policies (i.e. feedforward only). Whereas infrastructure code should always implement the full API (below), it would be annoying for short scripts and notebooks using non-stateful policies.

@naeioi It applies to all policies in the sense that it must be part of their function signature, but they are under no obligation to use it.

rollout example for general code:
```py
for _ in range(num_episodes):
    ps = p.get_reset_state()
    o = env.reset()

    done = False
    t = 0
    while not done and t < max_path_length:
        a, ps = policy.get_action(o, policy_state=ps)
        o, r, done, env_info = env.step(a)
```

rollout example for known-stateless code:
```python
for _ in range(num_episodes):
    o = env.reset()

    done = False
    t = 0
    while not done and t < max_path_length:
        a, _ = policy.get_action(o)
        o, r, done, env_info = env.step(a)

Reply:
Essentially, the goal of this API is to make policy modules stateless, and to force the users of those stateful modules to maintain the state themselves.

This avoids forcing stateful policy authors to maintain state between calls to `get_action`, which can make things like serialization and cloning fraught. For instance, the current TF RNNs maintain the RNN state internally, and assume that sequential calls to `get_actions` (between calls to `reset`) will always have the same batch size. The policy now has an implicit internal state machine, which the caller has to mentally-model themselves.

I agree this problem is thorny, and I'm open to hearing different perspectives on how we could solve it.

The current practice seems reasonable on the face of things, but it can get messy:
* If you forget to call `reset()`, because you designed your code for stateless policies, your policy isn't actually stateful in the way you want it to be. This API makes it clear that if you want state, you will have to keep it yourself. 
* Under this API, passing `policy_state=None` to an RNN policy will lead to a `ValueError`, whereas right now you can just happily call `get_actions` repeatedly and always be accumulating state improperly. Right now, we could raise a `RuntimeError` if you forget to call `reset()` the first time, but after that we have no idea if you should have `reset()` or not.
* RNN policies have slightly-different APIs during optimization, so RNN support has to be added as a special case in each algorithm. Unifying the API allows us to add stateful policy support right out of the gate.
* You can't clone/pickle+unpickle a stateful policy with one batch size, maintaining its internal state, and sample it with another batch size (e.g. 1) without resetting the state. For instance, this could make evaluation sampling difficult, because the optimization target policy usually has a large batch size, and the evaluation policy usually a small one.

Reply:
Okay, I'm convinced this API is about as good as we can do.